### ISSUE TITLE
make weavedns work with `--net=none`

### ISSUE THREAD

**Issue by rade** [weaveworks/weave#1301]
weave networking works just fine when containers are started with `--net=none`, but weavedns doesn't:
````
$ weave launch
$ weave run --name=foo -ti ubuntu
6102b1f5fb73351f5f0bc1d625830833c524db12d51b2a0e1bd76e87353a39ff
$ weave run --name=bar -ti ubuntu
445978b5c76e3b80529cf081943370ed0ac99e3d20a6c8749d4c708703e00a5d
$ weave run --name=baz --net=none -ti ubuntu
b006487dcddd474115b1bb396039b3d7fae4ba700fb92f8cf5b29a50d6da12d4
$ docker exec bar ping -c 1 fooPING foo.weave.local (10.32.0.1) 56(84) bytes of data.
64 bytes from foo.weave.local (10.32.0.1): icmp_seq=1 ttl=64 time=0.052 ms

--- foo.weave.local ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.052/0.052/0.052/0.000 ms
$ docker exec baz ping -c 1 foo
ping: unknown host foo
$ docker exec baz ping -c 1 10.32.0.1
PING 10.32.0.1 (10.32.0.1) 56(84) bytes of data.
64 bytes from 10.32.0.1: icmp_seq=1 ttl=64 time=0.142 ms

--- 10.32.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.142/0.142/0.142/0.000 ms
$ 
````

We should do something about that.

**Comment by rade** [comment]
We could give the weave bridge and IP address and make weavedns listen on that. We can configure routes in the containers for that IP. The problem is routing the *reply*. We don't know a priori what subnet(s) containers might be placed it.

An alternative is to create a separate bridge and `ethwelcl` interface. Then we can give containers a "local" IP on that interface, on a known subnet.

**Comment by tomwilkie** [comment]
IIUC, this is because a container with --net=none doesn't get put on the docker bridge, and therefore can't talk to the dns server listening on the docker bridge ip

**Comment by dpw** [comment]
Have any users mentioned this as a problem?  Are there any real-world use-cases it prevents?  I note that with plain docker, `--net=none` means you get no DNS at all.  So maybe it's fine that weavedns does not work in that context.

**Comment by rade** [comment]
DNS via docker is used for resolving external names. By contrast weavedns is used for resolving names within a container network. It's a weave feature. It should not depend in Docker networking.

We know of some users who disable Docker networking with `--net=none`, yet want name resolution for containers to work.

**Comment by bboreham** [comment]
> I note that with plain docker, --net=none means you get no DNS at all.

It seems to make some attempt to give you DNS:

````
$ sudo docker run --net=none --dns=1.2.3.4 -ti ubuntu
root@2f883b97889f:/# cat /etc/resolv.conf 
nameserver 1.2.3.4
````

**Comment by squaremo** [comment]
When weave is used as a plugin, it will often be the case that the weave interface is the only interface that a container has. But then, few if any of those integrations will be bothered about using Weave's DNS.

**Comment by alex-sherwin** [comment]
Right now in docker stable this can be problematic with some containers, I'm running into this scenario now:

Trying to setup a ZooKeeper ensemble on a Weave network, which requires special handling to ensure that ZooKeeper binds to the ethwe interface IP and not the default docker bridge IP.  

I have to do this, because if it binds on 0:0:0:0, this particular application doesn't expose enough configurability to ensure that it's advertised addresses are the correct one (the weave interface)

But, if it only binds to the weave interface, then any exposed port mappings of the docker container do not work since it is not listening on the docker bridge interface either (which from what I gather, the docker-proxy requires to make the exposed ports to work)

I think all of this would be greatly simplified if it were an option to use --net=none, leaving just a ethwe interface..

But @squaremo said "When weave is used as a plugin, it will often be the case that the weave interface is the only interface that a container has".  I'm assuming this is only true for Docker unstable right now since networking hasn't been promoted to stable plugin status yet?

**Comment by hectorj2f** [comment]
Yes, we also have the same problem. We need to access to the `weavedns` but this one is on the Docker network, so it isn't reachable if you try to ping other hostnames in the weave network. Is there any way of having the `weavedns` on both the docker and weave network?. 
In such a way, we could reach the weavedns and keep using `--net=none`. But I am not sure if this solution could work out.

**Comment by rade** [comment]
> Is there any way of having the weavedns on both the docker and weave network?

There is no point in having it on both; having it on just the weave network would be sufficient. But that is hard, otherwise we would have done it that way in the first place.

**Comment by hectorj2f** [comment]
I just tried what I mentioned above in my local setup.

```
dns_ip=172.16.1.100
$ container_id=$(docker run --net=none -dti --hostname=tintim.weave.local --dns=$dns_ip --dns-search=weave.local ubuntu)
$ weave attach 172.16.1.69/16 $container_id
$ weave attach 172.16.1.100/16 weavedns
```

and I can then ping the other hostnames. But of course, I don't know of the implications of this setup :/.

**Comment by rade** [comment]
> I don't know of the implications of this setup :/.

It only works when weavedns and the app containers have IP addresses in the same sub-net. Which generally won't be the case.

**Comment by hectorj2f** [comment]
Thanks for the quick reply @rade. I see the restriction... 
Is it on your roadmap an existing solution or workaround to address this issue ? we want to use `--net=none` and still having name resolution for our containers.

**Comment by rade** [comment]
> Is it on your roadmap to address this issue

yes.

**Comment by hectorj2f** [comment]
Great ! @rade. Thanks.

**Comment by rade** [comment]
I wonder whether a combination of a route in the app containers (to get the DNS queries sent out of ethwe) and some iptables NAT'ing/masquerading on the host (to get the replies back) would work.

**Comment by bboreham** [comment]
Note similar requirement from user on IRC: https://botbot.me/freenode/weavenetwork/2015-12-09/?msg=55748385
